src/hyperiond-starter/configHandle.cpp

```diff
@@ -0,0 +1,274 @@
+			{
+				_process.stopProcessByCName(cName);
+			}
+		}
+	}
+}
+
+QString configHandle::convertToCName(QString pName)
+{
+	pName = pName.trimmed();
+	while(pName.startsWith("."))
+		pName.remove(0,1);
+
+	if(!pName.endsWith(".json"))
+		pName.append(".json");
+
+	return pName;
+}
+
+void configHandle::fileChanged(const QString path)
+{
+	processConfigChange(path);
+}
+
+//Snippet from : https://forum.qt.io/post/393109
+void configHandle::modifyJsonValue(QJsonValue& destValue, const QString& path, const QJsonValue& newValue)
+{
+	const int indexOfDot = path.indexOf('.');
+	const QString dotPropertyName = path.left(indexOfDot);
+	const QString dotSubPath = indexOfDot > 0 ? path.mid(indexOfDot + 1) : QString();
+
+	const int indexOfSquareBracketOpen = path.indexOf('[');
+	const int indexOfSquareBracketClose = path.indexOf(']');
+
+	const int arrayIndex = path.mid(indexOfSquareBracketOpen + 1, indexOfSquareBracketClose - indexOfSquareBracketOpen - 1).toInt();
+
+	const QString squareBracketPropertyName = path.left(indexOfSquareBracketOpen);
+	const QString squareBracketSubPath = indexOfSquareBracketClose > 0 ? (path.mid(indexOfSquareBracketClose + 1)[0] == '.' ? path.mid(indexOfSquareBracketClose + 2) : path.mid(indexOfSquareBracketClose + 1)) : QString();
+
+    // determine what is first in path. dot or bracket
+	bool useDot = (indexOfDot >= 0 && indexOfSquareBracketOpen >= 0) ? (indexOfDot <= indexOfSquareBracketOpen) : (indexOfSquareBracketOpen < 0);
+
+	QString usedPropertyName = useDot ? dotPropertyName : squareBracketPropertyName;
+	QString usedSubPath = useDot ? dotSubPath : squareBracketSubPath;
+
+	QJsonValue subValue;
+	if (destValue.isArray())
+	{
+		subValue = destValue.toArray()[usedPropertyName.toInt()];
+	}
+	else if (destValue.isObject())
+	{
+		subValue = destValue.toObject()[usedPropertyName];
+	}
+	else
+	{
+		qDebug() << "oh, what should i do now with the following value?! " << destValue;
+	}
+
+	if(usedSubPath.isEmpty())
+	{
+		subValue = newValue;
+	}
+	else
+	{
+		if (subValue.isArray())
+		{
+			QJsonArray arr = subValue.toArray();
+			QJsonValue arrEntry = arr[arrayIndex];
+			modifyJsonValue(arrEntry,usedSubPath,newValue);
+			arr[arrayIndex] = arrEntry;
+			subValue = arr;
+		}
+		else if (subValue.isObject())
+		{
+			modifyJsonValue(subValue,usedSubPath,newValue);
+		}
+		else
+		{
+			subValue = newValue;
+		}
+	}
+
+	if (destValue.isArray())
+	{
+		QJsonArray arr = destValue.toArray();
+		arr[arrayIndex] = subValue;
+		destValue = arr;
+	}
+	else if (destValue.isObject())
+	{
+		QJsonObject obj = destValue.toObject();
+		obj[usedPropertyName] = subValue;
+		destValue = obj;
+	}
+	else
+	{
+		destValue = newValue;
+	}
+}
+
+void configHandle::modifyJsonValue(QJsonDocument& doc, const QString& path, const QJsonValue& newValue)
+{
+	QJsonValue val;
+	val = doc.isArray() ? doc.array() : doc.object();
+
+	modifyJsonValue(val,path,newValue);
+
+	doc = QJsonDocument((val.isArray() ? val.toArray() : val.toObject() );
+}
```

Comment:
kompiliert leider nich. daher dann doch eher so:
```
  QJsonValue val;
  if (doc.isArray())
  {
    val = doc.array();
    modifyJsonValue(val,path,newValue);
    doc = QJsonDocument(val.toArray());
  }
  else
  {                                                                                                                                                                                               
    val = doc.object();
    modifyJsonValue(val,path,newValue);
    doc = QJsonDocument(val.toObject());
  }  ```

... allerdings versteh icg die rekursion darin noch nicht. Kannst mal sagen was du da vor hast? Mir fehlt da auch die abbruch bedingung, sodass das nich unendlich läuft

Reply:
Also, praktisch nicht mein code! Also hab ich auch nicht viel gedacht dabei :)

Reply:
Wenn ich einen Vorschlag machen dürfte. :-)
Die Funktion "modifyJsonValue" existiert bereits in der "QJsonUtils.h" unter dem FuncNamen "modify".

Damit würden wir Code Duplicates vermeiden.

**configHandle.h:**

`#include <utils/jsonschema/QJsonUtils.h>`

**configHandle.cpp:**

`bool configHandle::createConfig(QString configName, QString pName)`
`{`
`	// init default config`
`	Q_INIT_RESOURCE(resource);`

`	// get default config for edits`
`	QString config = readFile(":/hyperion_default.config");`

`	QJsonDocument doc = QJsonDocument::fromJson(config.toUtf8());`
`	QJsonObject obj = doc.object();`
`	quint16 jsonPort = obj["jsonServer"].toObject().take("port").toInt();`
`	quint16 protoPort = obj["protoServer"].toObject().take("port").toInt();`
`	quint16 webPort = obj["webConfig"].toObject().take("port").toInt();`

`	jsonPort = checkPort(jsonPort);`
`	protoPort = checkPort(protoPort);`
`	webPort = checkPort(webPort, true);`

`	_usedPorts << jsonPort << protoPort << webPort;`

**`	QJsonUtils::modify(obj, QStringList() << "jsonServer" << "port",  jsonPort);`**
**`	QJsonUtils::modify(obj, QStringList() << "protoServer" << "port",  jsonPort);`**
**`	QJsonUtils::modify(obj, QStringList() << "webConfig" << "port",  jsonPort);`**
**`	QJsonUtils::modify(obj, QStringList() << "general" << "port",  jsonPort);`**

**`	doc.setObject(obj);`**

`	// write the final doc`
`	QFile tFile(_configPath+configName);`
`	if(tFile.open(QFile::WriteOnly))`
`	{`
`		tFile.write(doc.toJson());`
`		tFile.close();`
`		qDebug() << "config creation success:" << configName;`
`		return true;`
`	}`
`	else`
`	{`
`		qDebug() << "config creation FAILED:" << configName;`
`		return false;`
`	}`
`}`